packages/@aws-cdk/aws-eks/README.md

```diff
@@ -811,6 +812,84 @@ const chart2 = cluster.addHelmChart(...);
 chart2.node.addDependency(chart1);
 ```
 
+#### CDK8s Charts
+
+[CDK8s](https://cdk8s.io/) is an open-source library that enables Kubernetes manifest authoring using familiar programming languages. It is founded on the same technologies as the AWS CDK such as [`constructs`](https://github.com/aws/constructs) and [`jsii`](https://github.com/aws/jsii).
+
+> To learn more about cdk8s, visit the [Getting Started](https://github.com/awslabs/cdk8s/tree/master/docs/getting-started) tutorials.
+
+The EKS module natively integrates with cdk8s and allows you to apply cdk8s charts on AWS EKS clusters via the `cluster.addCdk8sChart` method.
+
+To get started, add the `cdk8s` dependency to your `package.json` file.
+
+```json
+dependencies: {
+  "cdk8s": "^0.29.0"
+}
+```
+
+> Note that the version of `cdk8s` must be `>=0.29.0`.
+
+Next, install the `cdk8s-cli`:
+
+```console
+npm install -g cdk8s-cli
+```
+
+The CLI is only used to import the core kubernetes api objects, and is not necessarily required as a runtime dependency. In your project directory, run the following command:
+
+```console
+cdk8s import k8s@{kubernetes-cluster-version} -l typescript
+```
+
+For example, for `eks.KubernetesVersion.V1_17` you would run:
+
+```console
+cdk8s import k8s@1.17.0 -l typescript
+```
+
+This will create an `imports` directory that contains the entire k8s object API, as typescript constructs. We recommend you commit this directory as part of your source code.
+
+In your application code:
+
+```ts
+import * as eks from '@aws-cdk/aws-eks';
+import * as cdk8s from 'cdk8s';
+import * as k8s from './imports/k8s';
+
+const cluster = new eks.Cluster(this, 'HelloCDK8s', {
+  version: eks.KubernetesVersion.V1_17,
+});
+
+// create a cdk8s app
+const cdk8sApp = new cdk8s.App();
+
+// create a cdk8s chart
+const proxy = new cdk8s.Chart(cdk8sApp, 'ProxyChart');
```

Comment:
Remind me why we can just add the chart directly under the AWS CDK scope? They all derive from the base construct

Reply:
Unfortunately, this isn't currently possible. It ok as far as compilation, but fails at runtime.

The first problem is that the `chart.toJson()` method isn't isolated to the chart, but rather also uses the chart's `App`.

https://github.com/awslabs/cdk8s/blob/master/packages/cdk8s/src/app.ts#L78

```ts
const app: App = App.of(chart);

// we must prepare the entire app before synthesizing the chart
// because the dependency inference happens on the app level.
resolveDependencies(app);
```

Instantiating a chart with an `aws-cdk` construct as the scope, results in `App.of(chart)` to return the `aws-cdk.App` instance. This instance is then passed on to `resolveDependencies`:

```ts
for (const dep of Node.of(app).dependencies) {
  ...
  ...
  // create an explicit chart dependency from implicit construct dependencies
  const sourceChart = Chart.of(dep.source);
  const targetChart = Chart.of(dep.target);
}
``` 

This is where it breaks down, `Node.of(app).dependencies` now returns all the `aws-cdk` dependency relationships, which means that `Chart.of` is now invoked on `aws-cdk` constructs, resulting in the following error:

```console
Error: cannot find a parent chart (directly or indirectly)
    at Function.of (/path/to/src/node_modules/cdk8s/src/chart.ts:30:13)
    at Function.of (/path/to/src/node_modules/cdk8s/src/chart.ts:33:18)
    at Function.of (/path/to/src/node_modules/cdk8s/src/chart.ts:33:18)
    at Function.of (/path/to/src/node_modules/cdk8s/src/chart.ts:33:18)
    at Function.of (/path/to/src/node_modules/cdk8s/src/chart.ts:33:18)
    at resolveDependencies (/path/to/src/node_modules/cdk8s/src/app.ts:136:31)
    at Function._synthChart (/path/to/src/node_modules/cdk8s/src/app.ts:84:5)
    at Chart.toJson (/path/to/src/node_modules/cdk8s/src/chart.ts:85:16)
    at Cluster.addCdk8sChart (/path/to/src/packages/@aws-cdk/aws-eks/lib/cluster.ts:63
```

This problem is actually solvable, and I already have a [branch](https://github.com/awslabs/cdk8s/tree/epolon/isolated-chart-synth) with a possible fix, basically making it so that the chart synthesis is isolated from the app. Thats not to say this is necessarily the right solution.

The second problem, and the more acute, is the reverse.

When a `cdk8s` construct is added as a child of an `aws-cdk` construct, the `aws-cdk` synthesis mechanism eventually does:

https://github.com/aws/aws-cdk/blob/60c782fe173449ebf912f509de7db6df89985915/packages/%40aws-cdk/core/lib/private/synthesis.ts#L49-L57

Specifically, notice the call to `root.node.children`. The problem is that an `aws-cdk` construct is a `Construct` class from the [compatibility layer](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/construct-compat.ts) inside `@aws-cdk/core`, while a `cdk8s` construct is a `Construct` class directly from the [`constructs`](https://github.com/aws/constructs/blob/master/src/construct.ts) module. 

The former **has** a `node` property, while the latter **doesn't**, resulting in:

```console
ypeError: Cannot read property 'children' of undefined
    at synthNestedAssemblies (/path/to/src/packages/@aws-cdk/core/lib/private/synthesis.ts:51:33)
    at synthNestedAssemblies (/path/to/src/packages/@aws-cdk/core/lib/private/synthesis.ts:66:7)
    at synthNestedAssemblies (/path/to/src/packages/@aws-cdk/core/lib/private/synthesis.ts:66:7)
    at Object.synthesize (/path/to/src/packages/@aws-cdk/core/lib/private/synthesis.ts:15:3)
    at App.synth (/path/to/src/packages/@aws-cdk/core/lib/stage.ts:174:23)
    at synthesizeApp (/path/to/src/packages/@aws-cdk/assert/lib/synth-utils.ts:76:15)
    at Function._synthesizeWithNested (/path/to/src/packages/@aws-cdk/assert/lib/synth-utils.ts:53:22)
    at Object.expect (/path/to/src/packages/@aws-cdk/assert/lib/expect.ts:9:45)
    at Object.cdk8s chart can be added to cluster (/path/to/src/packages/@aws-cdk/aws-eks/test/test.cluster.ts:48:5)
    at Object.<anonymous> (/path/to/src/node_modules/nodeunit/lib/core.js:236:16)
```

I suspect there are a multitude of other places where `aws-cdk` relies on the scope being a construct from the compatibility layer rather the from `constructs.

Im not sure this is solvable before we rollout v2.

Reply:
@nija-at what do you think about this? I am curious whether the change you are making today has any impact on this.

Reply:
Unfortunately not. 
I believe the AWS CDK today, expects all constructs in the tree to be a CDK construct at synth time. I've not dived into how deep this goes, or what it would take to fix it.